LNX-132 Make ticking depend on get. Add CreateObject action when adding object

alpinus4 opened the pull request:
It is a necessary change for the multiplayer to work.
The frontend has to have disabled ticking functionality, and handle CreateObject actions.

alpinus4 committed:
Make ticking depend on get. Ticks every 1s.

filip-szweda reviewed (approved):
Looks good to me

alpinus4 committed:
Small fixes

EasyJB reviewed (approved):
For now looking good.

alpinus4 committed:
Fix problem with object not existing when we create process for it

alpinus4 committed:
dd pytest timeout, remove tick from test_spam_objects

alpinus4 committed:
Fix spam objects test, remove pytest timeout

The tests don't handle background tasks, therefore remove adding object
with tick from spam objects test.
Pytest report timeout as nonexistent parameter, remove it.

alpinus4 commented:
I removed adding object with tick parameter from spam objects test, as the test just wasn't working with fastapi backgroundTasks.

filip-szweda commented on src/app.py:
```diff
@@ -32,6 +34,7 @@ class GlobalState:
     processes: Dict = None,
     transitions: Dict = None,
     tick_number: int = 0
+    last_tick: float = 0
```
`last_tick: float = 0.0`

filip-szweda commented on src/app.py:
```diff
@@ -17,6 +18,7 @@
 from lynx.common.scene import Scene
 from lynx.common.vector import Vector
+from lynx.common.actions.create_object import CreateObject
 from pydantic import BaseModel
```
Separate standard library, imported library and our own library imports if you have a chance

filip-szweda commented on src/app.py:
```diff
@@ -100,17 +103,43 @@ async def send_scene():
 
     await asyncio.gather(*future_sends)
 
+
```
I think we should commit reformatting changes separately, this shouldn't be here

filip-szweda commented on src/app.py:
```diff
@@ -100,17 +103,43 @@ async def send_scene():
+
+async def wait_for_next_tick():
+    tick = state.tick_number
+    while state.tick_number != tick + 1:
```
```
next_tick_number = state.tick_number + 1
while state.tick_number != next_tick_number:
```

filip-szweda commented on src/app.py:
```diff
@@ -100,17 +103,43 @@ async def send_scene():
+async def wait_for_next_tick():
+    tick = state.tick_number
+    while state.tick_number != tick + 1:
+        await asyncio.sleep(0.1)
```
Why `0.1` specifically?

alpinus4 replied:
Just a small time amount, no particular reason

filip-szweda commented on src/app.py:
```diff
@@ -100,17 +103,43 @@ async def send_scene():
+async def spawn_process_for_new_agent(object: Object):
+    await wait_for_next_tick()
+
+    parent_conn, child_conn = AioPipe()
```
Please expand on what `conn` is

alpinus4 replied:
> `parent_conn` and `child_conn` refer to the parent and child connections, respectively, in a multiprocessing setup. These connections are typically used for interprocess communication (IPC) between a parent process and a child process.

> In the code snippet you provided, `AioPipe()` is used to create a pair of asynchronous pipes for communication between the parent and child processes. The `AioPipe()` function likely creates two connected `asyncio` streams, one for the parent process (`parent_conn`) and one for the child process (`child_conn`).

> With these pipes, you can send messages or data between the parent and child processes asynchronously. The parent process can write data to the `parent_conn` pipe, and the child process can read from it using the `child_conn` pipe, and vice versa.

> This setup is commonly used in multiprocessing scenarios where you need to distribute tasks among multiple processes and exchange data or messages between them.

filip-szweda commented on src/app.py:
```diff
@@ -100,17 +103,43 @@ async def send_scene():
+    await wait_for_next_tick()
+
+    parent_conn, child_conn = AioPipe()
+    p = AioProcess(target=execution_runtime,
```
`process`, we're cool with longer variable names

filip-szweda commented on src/app.py:
```diff
@@ -100,17 +103,43 @@ async def send_scene():
+
+    parent_conn, child_conn = AioPipe()
+    p = AioProcess(target=execution_runtime,
+                    args=(child_conn, object.id,))
```
Should be inline with `p = AioProcess(target=execution_runtime,`

filip-szweda commented on src/app.py:
```diff
@@ -100,17 +103,43 @@ async def send_scene():
+
+    await parent_conn.coro_send(state.scene.serialize())
+    state.processes[object.id] = ProcessData(
+        process=p, pipe=parent_conn)
```
Should be inline with `state.processes[object.id] = ProcessData(`
I think you should set longer line breaks in IDE

blazej-smorawski commented on tests/scene_server_test.py:
```diff
@@ -42,14 +42,11 @@ async def test_spam_objects(self):
                 response = await TestSceneServer.fetch(ac, "/?tick_number=0")
-                # scene = Scene.deserialize(response['scene'])
+
+                # we cannot test add_object with tick here, AsyncManager doesn't handle BackgroundTasks
```
I wouldn't say it solves the problem, I think that we need such test. We could add some smaller tests if this one is not possible too.

alpinus4 committed:
Review fixes

filip-szweda commented on pytest.ini:
```diff
@@ -1,2 +1,4 @@
 [pytest]
-pythonpath = .
\ No newline at end of file
+pythonpath = .
```
Is it safe to use "."?

alpinus4 replied:
I made my additions consistent with what was there before

filip-szweda commented on pytest.ini:
```diff
@@ -1,2 +1,4 @@
 [pytest]
-pythonpath = .
\ No newline at end of file
+pythonpath = .
```
Why spaces before and after "="?

filip-szweda commented on src/app.py:
```diff
@@ -33,6 +36,7 @@ class GlobalState:
     processes: Dict = None,
     transitions: Dict = None,
     tick_number: int = 0
+    last_tick: float = 0.0
```
`last_tick_time`

filip-szweda commented on src/app.py:
```diff
@@ -109,18 +113,41 @@ async def send_scene():
+    await wait_for_next_tick()
+
+    parent_connection, child_connection = AioPipe()
+    process = AioProcess(target=execution_runtime,
```
Why is the code broken between two lines?

filip-szweda commented on src/app.py:
```diff
@@ -109,18 +113,41 @@ async def send_scene():
+    while state.tick_number != next_tick_number:
+        await asyncio.sleep(0.1)
+
+async def spawn_process_for_new_agent(object: Object):
```
`-> NoReturn`

alpinus4 replied:
Let's add `NoReturn` to functions in separate refactor PR

filip-szweda commented on src/app.py:
```diff
@@ -130,20 +157,11 @@ async def get(tick_number: int) -> Dict[str, Union[int, str, List[List[str]]]]:
 
 @app.post("/add_object")
-async def add(r: AddObjectRequest):
+async def add(r: AddObjectRequest, background_tasks: BackgroundTasks):
```
`-> NoReturn`

alpinus4 replied:
As before

filip-szweda commented on src/app.py:
```diff
@@ -130,20 +157,11 @@ async def get(tick_number: int) -> Dict[str, Union[int, str, List[List[str]]]]:
     object = Object.deserialize(r.serialized_object)
-    state.scene.add_entity(object)
+    state.scene.add_to_pending_actions(CreateObject(r.serialized_object).serialize())
     if object.tick != "":
```
Would be cool to have a comment here explaining that we only spawn processes for agents with code
Also, we really need to change the name of this attribute (`tick`, I mean) in the future xD

filip-szweda reviewed (changes requested):
The logic itself looks good

alpinus4 committed:
Code style changes

filip-szweda commented on pytest.ini:
```diff
@@ -1,4 +1,4 @@
-log_cli=true  
-log_level=INFO
\ No newline at end of file
+log_cli = true  
```
Thought that "=" is better than " = "? A whatever, but made me thinking

alpinus4 replied:
I like it more with spaces, it just looks better.
It depends, in some places it matters, in others it doesn't.
If I can choose I would go for a variant with spaces any day.

filip-szweda reviewed (approved):
'Tis enough
